packages/core/src/sync/service.ts

```diff
@@ -142,9 +155,28 @@ export const create = async ({
       }
 
       case "finalize": {
-        syncService.networkServices.find(
+        const networkService = syncService.networkServices.find(
           (ns) => ns.network.chainId === realtimeSyncEvent.chainId,
-        )!.realtime!.finalizedCheckpoint = realtimeSyncEvent.checkpoint;
+        )!;
+
+        // "realtime" property may be undefined when `kill()` has been
+        // invoked but hasn't completed.
+        if (networkService.realtime === undefined) return;
+
+        networkService.realtime!.finalizedCheckpoint =
+          realtimeSyncEvent.checkpoint;
+
+        // Check if the finalized blockNumber is greater than the end block of all
+        // sources for the network. Potentially kill the realtime sync and remove the
+        // network from checkpoint calculations.
+        if (
+          networkService.realtime.endBlock !== undefined &&
+          realtimeSyncEvent.checkpoint.blockNumber >
+            networkService.realtime.endBlock
+        ) {
+          networkService.realtime.realtimeSync.kill();
```

Comment:
I think we should add an info log here: `Synced final end block for '${network.name}' (${networkService.realtime.endBlock}), killing realtime sync service`